Explore: fixes toolbars datasource selector and date picker responsiveness

Estrax opened the pull request:
**What this PR does / why we need it**:

Adds responsive styles for datasource selector and updates responsiveness of date picker with long dates.

**Which issue(s) this PR fixes**:

Fixes #19353 
Fixes #19351

Commit:
Explore: fixes toolbars datasource selector and date picker responsiveness

Commit:
Explore: updates grafana UI time picker component - adds a class on long date

Commit:
Explore: updates styles for responsive long date without split

Commit:
Explore: adds styles for responsive time picker with long date during split

Commit:
Explore: updates time picker styles

Estrax commented:
@kaydelaney @davkal wdyt?

davkal reviewed (dismissed):
Works well and could go in as is.

davkal commented on packages/grafana-ui/src/components/TimePicker/TimePicker.tsx:
```diff
@@ -183,6 +183,7 @@ class UnThemedTimePicker extends PureComponent<Props, State> {
       </>
     );
     const isAbsolute = isDateTime(value.raw.to);
+    const isLongDate = typeof this.props.value.raw.from !== 'string' || typeof this.props.value.raw.to !== 'string';
```
Wouldnt `isAbsolute` be enough?

Estrax replied:
@davkal Sure, though we are going to have small inconsistencies then.

> 2019-10-01 00:00:00 to a few seconds ago
aka **date-to-now** would be a short date (with isAbsolute).

> 5 minutes ago to 2019-10-09 23:59:59
aka **now-minus-offset-to-date** would be a long date, even though technically it's shorter than date-to-now.

With **isLongDate** both of them are classified as long dates.

Estrax replied:
@davkal Sure, though we are going to have small inconsistencies then.

With isAbsolute:

> 2019-10-01 00:00:00 to a few seconds ago

aka **date-to-now** would be classified as a short date.

> 5 minutes ago to 2019-10-09 23:59:59

aka **now-minus-offset-to-date** would be classified a long date, even though technically it's shorter than date-to-now in an example above.

With **isLongDate** both of them are classified as long dates.

davkal replied:
I think for the purpose of collapsing the button, it's fine if either is already an absolute date, perhaps `isAbsolute` could change to `hasAbsolute` if that works with the rest of the code. 

I guess the biggest issue I have is that we're doing 2 different kinds of detection here: one with `typeof` and the other one with `isDateTime` just above. Can we use either and not both? If we need to use both, then I'd expect a comment.

Estrax replied:
@davkal of course, we don't need them both. I already fixed that - renamed isAbsolute to hasAbsolute and consolidated long datetime detection within hasAbsolute.

```ts
const hasAbsolute = isDateTime(value.raw.from) || isDateTime(value.raw.to);
```

davkal replied:
Happy with this.

davkal commented:
@aocenas would that have been a good use for the `ResponsiveButton`?

aocenas commented:
Should the datasource picker also be responsive? At the moment this is what I see on tablet size:
<img width="764" alt="Screenshot 2019-10-09 at 10 23 07" src="https://user-images.githubusercontent.com/1014802/66464292-fcd35d00-ea7e-11e9-9676-1c19da9b6154.png">

> would that have been a good use for the ResponsiveButton?

Probably yes. Though not sure how TimePicker is done atm and whether it will be easy to just inject it there. Also as I think TimePicker is one that is used everywhere else it will make it responsive everywhere.

As a side note, I do not particularly like the approach of using just the icons. For things like run button it would make sense as the `Run` text does not change but with TimePicker you loose the info of what range we are showing and I doubt people know datasources by icons. Personally I would rather have it just stacked so it does not overflow, which is uglier but more functional and maybe use some more compact date format.

Commit:
Explore: updates long datetime detector to isDateTime, renames isAbsolute to hasAbsolute

Commit:
Explore: merge changes

Commit:
Explore: merge changes

Commit:
Explore: updates datasource responsiveness

Estrax commented:
@davkal Fixed.

Normal view:
![Screenshot from 2019-10-12 03-04-41](https://user-images.githubusercontent.com/6322797/66692276-21a51b80-ec9d-11e9-90b1-bff5859e3431.png)

Split view:
![Screenshot from 2019-10-12 03-04-26](https://user-images.githubusercontent.com/6322797/66692304-6761e400-ec9d-11e9-8bdf-d474f6ba03cd.png)

Select open:
![Screenshot from 2019-10-12 03-12-38](https://user-images.githubusercontent.com/6322797/66692388-3d5cf180-ec9e-11e9-83cb-b6e83422f749.png)

Requires a modified version of @torkelo/react-select, so it's going to be mergeable right when the fix is going to be published (very soon).

aocenas commented:
@Estrax is the small version of the select going to be only in Explore? Because I assume this won't work with selects where values do not have image.

Estrax commented:
@aocenas yes (unless you add `explore-toolbar` class somewhere else)

torkelo commented on public/sass/pages/_explore.scss:
```diff
@@ -120,6 +128,33 @@
+          min-width: 50px;
+          max-width: 50px;
+
+          .gf-form-select-box__single-value {
```
I am not sure I think overriding internal css classes from explore is ideal from a maintainable standpoint. Feels a bit fragile. 

Can this be done via prop option on the on the select component instead?

Estrax replied:
Not sure if the react-select has the prop indicating the split, but I can refactor it to ResponsiveButton.

torkelo commented on public/sass/pages/_explore.scss:
```diff
@@ -148,10 +183,29 @@
+      min-width: 50px;
+      max-width: 50px;
+
+      .gf-form-select-box__single-value {
```
This is an internal class, style defined in _Select.scss  , overriding it from the outside like this is not ok.

Estrax replied:
Moved to _Select.scss and refactored it to be a prop.

torkelo commented on public/sass/pages/_explore.scss:
```diff
@@ -120,6 +128,33 @@
+        display: none;
+      }
+    }
+  }
```
The TimePicker is a generic component that is used from Dashboard (and potentially other places) as well, this should be in _TimePicker.scss

Estrax replied:
Moved to _TimePicker.scss.

torkelo commented on public/sass/pages/_explore.scss:
```diff
@@ -61,6 +61,14 @@
+        display: none;
+      }
+    }
+  }
```
The TimePicker is a generic component that is used from Dashboard (and potentially other places) as well, this should be in _TimePicker.scss

Estrax replied:
Moved to _TimePicker.scss.

torkelo reviewed (changes requested):
This has some css challenges that might be tricky to address. Overriding internal css classes for generic components from the outside is very fragile and would like to find other options.

Commit:
Explore: moves time picker styles

Commit:
Explore: updates datasource picker and select component responsiveness

Estrax commented:
@torkelo moved these styles to appropriate files, rewrote the collapse and made it a prop. Is it a little bit better now?

torkelo reviewed (commented):
Looks much better now, but still some things that do not feel right. 

the .datasource-picker class is a bit strange, should really be named .explore-ds-picker (and be defined in explore.css) as it not part of the data source picker. 

Then I think the ds-picker (the DataSourcePicker component) should have a class/property to make it with: 100% so you can control width from the outside, then the style rules on .datasource-picker does not need to override the internal class of DataSourcePicker. 

The collapse handling is also a bit off. It's about hiding the value, wondering if there is cleaner way to hide that value

Commit:
Explore: merge changes

torkelo reviewed (commented):
Pushed 2 commits to show what I mean: 
Using a new prop on the value that hides the text (instead of isCollapsed on the component) 
Also using containerWidth from redux state
https://github.com/grafana/grafana/tree/hide-text-in-ds-picker 

It needs more work (especially the use of containerWidth) as it does not match the media queries for the width of the component

Commit:
Explore: merge changes

Commit:
Explore: updates data source picker

Commit:
Explore: updates explore toolbar to use container width

Commit:
Explore: updates styles for datasource picker

Commit:
Explore: updates Grafana UI elements - select and single value with an ability to hide text

Commit:
Explore: updates time picker styles

Commit:
Explore: merge changes

torkelo commented on packages/grafana-ui/src/components/Select/Select.tsx:
```diff
@@ -50,6 +50,7 @@ export interface CommonProps<T> {
   onOpenMenu?: () => void;
   onCloseMenu?: () => void;
   tabSelectsValue?: boolean;
+  hideTextValue?: boolean;
```
this property on Select does not look like it is needed, can't see that it is used anywhere.

Estrax replied:
Right now it's being controlled by ExploreToolbar and passed down as a prop so it prevents hiding text from other Select components as they may use different breakpoints.

torkelo replied:
but cannot see it being used anywhere, it's set on the SelectComponent (which does not have any such property)

Estrax replied:
Ah, you're absolutely right. My bad. I missed to remove it from there as I forwarded this prop via value object.

torkelo commented on packages/grafana-ui/src/components/TimePicker/_TimePicker.scss:
```diff
@@ -250,6 +251,24 @@ $arrowPadding: $arrowPaddingToBorder * 3;
+
+  .time-picker-long-date {
+    .select-button {
+      span {
```
now we only have this internal override on an internal span element that belongs to the ButtonSelect component. So this time-picker-long-date class should be a property or BEM modifier class on the ButtonSelect not on the TimePicker component.

torkelo reviewed (commented):
Great updates, we are getting close. Still have the TimePicker scss overriding an internal element type that belongs to the ButtonSelect component.

torkelo commented on packages/grafana-ui/src/components/Select/Select.tsx:
```diff
@@ -251,6 +256,7 @@ export class AsyncSelect<T> extends PureComponent<AsyncProps<T>> {
               }}
               defaultValue={defaultValue}
               value={value}
+              hideText={hideTextValue}
```
this is the one that I don't think is used, as there is no such property on this component.

torkelo commented on packages/grafana-ui/src/components/Select/_Select.scss:
```diff
@@ -217,3 +220,7 @@ $select-input-bg-disabled: $input-bg-disabled;
   display: flex;
 }
+
+.gf-form-select-box__img-value-hidden {
```
where is this class set?

Estrax replied:
It was added when I was experimenting with pure CSS solution only. Did the same thing the other way, already cleaned it up.

Commit:
Explore: updates datasource select and date picker

torkelo reviewed (dismissed):
Code wise this looks good now! Thanks keeping up with all my change requests! 

@davkal can you do final test?

davkal commented:
Works well! Two last nits:

- in the non-split mode, the Explore is butted against the selector, would be great if the spaces between `Explore`, the selector, and the next button could be the same as the space between the rest of the toolbar buttons
<img width="425" alt="Screenshot 2019-10-25 at 15 53 03" src="https://user-images.githubusercontent.com/859729/67577220-62427180-f740-11e9-80d4-d4c2d6502375.png">

- datasources with long names look broken when they have mode switcher, e.g., Elastic, and the run button wraps to next line

<img width="1325" alt="Screenshot 2019-10-25 at 15 57 11" src="https://user-images.githubusercontent.com/859729/67577251-69697f80-f740-11e9-848d-ef1d36912965.png">

Commit:
Explore: Merge changes

Commit:
Explore: updates toolbar elements margin

Commit:
Explore: updates toolbar elements split breakpoints

Commit:
Explore: updates datasource picker label length with substrings

Commit:
Explore: updates the datasource picker label length

Estrax commented:
> * in the non-split mode, the Explore is butted against the selector, would be great if the spaces between `Explore`, the selector, and the next button could be the same as the space between the rest of the toolbar buttons

Fixed.

> * datasources with long names look broken when they have mode switcher, e.g., Elastic, and the run button wraps to next line

Fixed - set up max-width for this picker and cut the displayed label for a currently selected option to up to 36 characters. Full labels available (with a tick indicating a selected source) within a dropdown when the picker is open.

Commit:
Merge changes

davkal commented:
Run button disappears on smaller screens (it's fine if it wraps on really small screens like this here):

Loki (live button stays for some reason):
![Screenshot 2019-10-28 at 10 53 17](https://user-images.githubusercontent.com/859729/67673509-2989de80-f972-11e9-8711-e2588261bf94.png)

Elastic:
![Screenshot 2019-10-28 at 10 53 40](https://user-images.githubusercontent.com/859729/67673521-2e4e9280-f972-11e9-9bfe-6f6499c13e8b.png)

Commit:
Explore: removes refresh picker hide media query

Commit:
Explore: updates refresh picker style query to use xs breakpoint

Estrax commented:
> Run button disappears on smaller screens (it's fine if it wraps on really small screens like this here):
> 
> Loki (live button stays for some reason):
> ![Screenshot 2019-10-28 at 10 53 17](https://user-images.githubusercontent.com/859729/67673509-2989de80-f972-11e9-8711-e2588261bf94.png)
> 
> Elastic:
> ![Screenshot 2019-10-28 at 10 53 40](https://user-images.githubusercontent.com/859729/67673521-2e4e9280-f972-11e9-9bfe-6f6499c13e8b.png)

Solved. There was a media query hiding refresh picker below the `sm` breakpoint, changed it to `xs`.

Estrax commented:
Solved. There was a media query hiding refresh picker below the sm breakpoint, changed it to xs.

davkal commented on public/app/core/components/Select/DataSourcePicker.tsx:
```diff
@@ -45,10 +46,11 @@ export class DataSourcePicker extends PureComponent<Props> {
 
     const value = current && {
-      label: current.name,
+      label: current.name.substr(0, 37),
```
A bit harsh, but I like it.

davkal reviewed (commented):
Works well and looks good!

Nits: A couple of magic numbers in there that could either be extracted or use comments to tell what the goal of those widths is.
